File: algorithms/bubble.java
import java.io.*;
import java.util.*;

class bubble
{
	public static void main(String argvs[])
	{
	
		// define an array:
		// int len = 100000;
		// double[] arr = new double[len];

		// long start = System.currentTimeMillis();
		// //random number:
		// generatedRandom(arr);
		// long end = System.currentTimeMillis();
		// System.out.println((end-start)/1000L);

		// // dop the operations:
		// System.out.println("BUBBLE: ");
		// long start2 = System.currentTimeMillis();
		// bubbleSort(arr, len);
		// long end2 = System.currentTimeMillis();
		// System.out.println((end2-start2)/1000L);
		// System.out.println("Soted array: ");
		// printArray(arr, len);
	}

// 	//pass array as parameter:
// 	static void bubbleSort(double[] arr, int n)
// 	{
// 		boolean swap;
// 		for(int i=0;i<n-1;i++){

// 			swap = false;

// 			for(int j=0;j<n-1;j++){

// 				if(arr[j] > arr[j+1]){

// 					double tmp = arr[j];
// 					arr[j] = arr[j+1];
// 					arr[j+1] = tmp;

// 					swap = true;
// 				}
// 			}
// 			if(swap==false)
// 				break;
// 		}
// 	}

// 	static void generatedRandom(double[] arr)
// 	{
// 		for(int i=0;i<arr.length;i++){
// 			Random rn = new Random();
// 			double min = 10;
// 			double max = 20;
// 			arr[i] = min + rn.nextDouble() * (max - min);
// 		}
// 	}
// 	//pass array as parameter:
// 	static void printArray(double[] arr, int n)
// 	{
// 		for(int i=0;i<n;i++){
// 			System.out.print(arr[i]+" ");
// 		}
// 		System.out.println();
// 	}

}

File: algorithms/bubble.cpp
#include<stdio.h>
#include<iostream>
#include<cstdlib> //rand
#include<ctime>
using namespace std;

float* generateRandom();
void swap(float*, float*);
void bubble(float*, int);
void printArray(float* ,int);  //> (1)so here is a pointer float*
  
int length = 100000;
int main(int argc, char** argv)
{
	clock_t start,start2;
	double duration, duration2;
	
	cout<<"GENERATING...\n";
	start = clock();
	float* ptr = generateRandom(); 
	duration = (clock() - start)/(double) CLOCKS_PER_SEC;
	cout<<duration<<endl;

	cout<<"BUBBLE: \n";
	start2 = clock();
	bubble(ptr, length);
	duration2 = (clock() - start2)/(double) CLOCKS_PER_SEC;
	cout<<duration2<<endl;
	//printArray(ptr, length);

}

void bubble(float arr[], int n)  //
{
	for(int i=0;i<n;i++){

		double flag=false;
		for(int j=i+1;j<n;j++){

			if(arr[i]>arr[j]){
				swap(arr[i], arr[j]);

				flag = true;
			}
		}
		if(flag == false)
			break;
	}
}

float* generateRandom()
{
	float* ptr = new float[length];
	for(int i=0;i<length;i++){
		// (2) random int
		//floatNumber[i] = 1 + (rand() % static_cast<int>(100 - 1 + 1));
		// (2) random float form 0~10:
		*(ptr+i) = static_cast <float> (rand()) / (static_cast <float> (RAND_MAX/10));
		//cout<<ptr[i]<<" ";
	}
	return ptr;

	delete ptr;
	ptr=NULL;
}

void swap(float* px, float*py)
{
	float tmp = *px;
	*px = *py;
	*py = tmp;
}
void printArray(float arr[], int n)
{
	for(int i=0;i<n;i++){
		cout<<arr[i]<<" ";
	}
	cout<<"\n";
}
